### PR TITLE
[WIP] Expose EC APIs

### PR DESCRIPTION
**Pull request by benr-ml**
## Description 
Add Move modules for Ristretto and BLS12-381 group operations.
**[Gas related code is not yet ready, but the rest is in]**

## Test Plan 
Existing and many newly added tests.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [x] breaking change for FNs (FN binary must upgrade)
- [x] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
Add Move modules for Ristretto and BLS12-381 group operations (allowing developers to implement cryptographic protocols directly in Move).

### CODE REVIEW

**Comment by benr-ml** [comment]
@kchalkias @joyqvq @jonas-lj @arnab-roy Please take a look on the Move APIs and let me know if there is anything else you think we should add. Also, see if you have better ideas re the Move types -> Move native calls -> Rust interfaces (I chose the safest option for developers, but it required dispatching calls in the Rust side).

The two bigs parts I haven't finished yet:
- Examples - will add a few to this PR, and probably more in a follow up PR
- Fees - I'll fix those once we agree on the rest (mostly boilerplate code)

**Comment by benr-ml** [comment]
Added examples

**Comment by benr-ml** [comment]
@amnn @damirka This PR adds group operations APIs to Move. It doesn't have yet the gas related code, but the rest is here.

I'd appreciate your feedback on two things:
- The Move -> Rust flow (see `group_ops.rs`, `group_ops.move` and `ristretto255.move` for example) - I decided to use explicit functions per group, and dispatch internally to a small number of native functions. Let me know if you see a better alternative, or any issues with the current approach.
- `ec_ops.move` has several examples of real cryptographic primitives we can implement with those new APIs. Please take a look and let me know if there are patterns you find "ugly" or too expensive, and might worth more thinking (e.g., add a native function for xoring 2 `vector<u8>`?).

File: crates/sui-framework/packages/sui-framework/sources/crypto/bls12381.move [hunk 1]
```diff
@@ -18,4 +21,222 @@ module sui::bls12381 {
...
+    }
+
+    public fun g1_neg(e: &Element<G1>): Element<G1> {
+        g1_sub(&g1_identity(), e)
+    }
+
+    // Hash using DST = BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_
+    public fun hash_to_g1(m: &vector<u8>): Element<G1> {
+        group_ops::hash_to(G1_TYPE, m)
+    }
```
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/crypto/bls12381.move:145]: One way to potentially clean this up is to turn these into native functions that are generic on the element type, e.g.:

```move
    public native fun from_bytes<G>(bytes: &vector<u8>): Element<G>;
    public native fun identity<G>(): Element<G>;
    public native fun generator<G>(): Element<G>;
    public native fun add<G>(e1: &Element<G>, e2: &Element<G>): Element<G>;
    public native fun sub<G>(e1: &Element<G>, e2: &Element<G>): Element<G>;
    public native fun mul<G>(e1: &Element<Scalar>, e2: &Element<G>): Element<G>;
    public native fun div<G>(e1: &Element<Scalar>, e2: &Element<G>): Element<G>;

    public fun neg<G>(e: &Element<G>): Element<G> {
        sub(&identity(), e)
    }

    public native fun hash_to<G>(m: &vector<u8>): Element<G>;
```

Then in the native function implementation, you dispatch the implementation based on the type parameter (particularly important for `identity` and `generator` whose bytes will presumably differ based on which group type you are using.

**Review by amnn — COMMENTED** [review]
@damirka is the expert on efficient Move, so I will leave him to comment there, but in terms of interactions between native functions and Move, I've left a suggestion for a way to do this with less repetition on the Move side.

File: crates/sui-framework/packages/sui-framework/sources/crypto/bls12381.move (hunk 1 above)
Comment (benr-ml) [reply crates/sui-framework/packages/sui-framework/sources/crypto/bls12381.move:145]: Thanks for this proposal. The disadvantage with this option was that some user mistakes are only detectable in runtime (e.g., if I call `div` with Scalar type that does not work with G, or if I just call any of the functions with an unsupported type).

**Pull request closed without merging**